File: common/src/main/java/me/wyne/wutils/common/VectorUtils.java
package me.wyne.wutils.common;

import org.bukkit.util.Vector;

public final class VectorUtils {

    public static final Vector ZERO_VECTOR = new Vector();

    public static Vector getVector(String string, Vector def) {
        var args = new Args(string, ",");
        var x = args.get(0).isEmpty() ? def.getX() : Double.parseDouble(args.get(0));
        var y = args.get(1).isEmpty() ? def.getY() : Double.parseDouble(args.get(1));
        var z = args.get(2).isEmpty() ? def.getZ() : Double.parseDouble(args.get(2));
        return new Vector(x, y, z);
    }

    public static Vector getVectorOrZero(String string) {
        return getVector(string, ZERO_VECTOR);
    }

    public static String toString(Vector vector) {
        return vector.getX() + "," + vector.getY() + "," + vector.getZ();
    }

}

File: commonKt/src/main/kotlin/me/wyne/wutils/common/kotlin/vector/VectorUtils.kt
package me.wyne.wutils.common.kotlin.vector

import me.wyne.wutils.common.vector.VectorUtils
import org.bukkit.block.BlockFace
import org.bukkit.util.Vector

val Vector.ZERO: Vector
    get() = Vector()

fun Vector.getMin(vector: Vector): Vector =
    VectorUtils.getMin(this, vector)

fun Vector.getMax(vector: Vector): Vector =
    VectorUtils.getMax(this, vector)

fun Vector.isEmpty() =
    VectorUtils.isEmpty(this)

fun Vector.addRelative(horizontal: Double, vertical: Double, face: BlockFace) =
    VectorUtils.addRelative(this, horizontal, vertical, face)

fun Vector.addRelative(width: Double, height: Double, depth: Double, face: BlockFace) =
    VectorUtils.addRelative(this, width, height, depth, face)

fun Vector.addRelative(relativeOffset: Vector, face: BlockFace) =
    VectorUtils.addRelative(this, relativeOffset, face)

fun Vector.addRelative(relativeOffset: Vector, forward: Vector) =
    VectorUtils.addRelative(this, relativeOffset, forward)

